Sources/Storage.swift

```diff
@@ -8,28 +8,117 @@
 
 import Foundation
 
-open class Storage: NSObject {
+open class Storage {
+    public enum Change {
+        case appended([RequestModel])
+        case removed(indice: Range<Int>)
+        case updated(RequestModel, at: Int)
+        case cleared
+    }
+
+    public final class Token: Hashable {
+        weak var storage: Storage?
+
+        fileprivate init(storage: Storage) {
+            self.storage = storage
+        }
+
+        public func stopObservation() {
+            storage?.transaction { storage?.observers[self] = nil }
+        }
+
+        deinit {
+            stopObservation()
+        }
+
+        public static func == (lhs: Token, rhs: Token) -> Bool {
+            return lhs === rhs
+        }
+
+        public func hash(into hasher: inout Hasher) {
+            hasher.combine(ObjectIdentifier(self))
+        }
+    }
 
     public static let shared: Storage = Storage()
     
-    open var requests: [RequestModel] = []
-    
-    func saveRequest(request: RequestModel?){
-        guard request != nil else {
-            return
+    open private(set) var requests: [RequestModel] = []
+    private var observers: [Token: (Change) -> Void] = [:]
+    private let lock: os_unfair_lock_t
+
+    init() {
+        lock = os_unfair_lock_t.allocate(capacity: 1)
+        lock.initialize(to: os_unfair_lock_s())
+
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(observeMemoryWarning),
+            name: UIApplication.didReceiveMemoryWarningNotification,
+            object: nil
+        )
+    }
+
+    deinit {
+        lock.deinitialize(count: 1)
+        lock.deallocate()
+    }
+
+    func observe(_ action: @escaping (Change) -> Void) -> Token {
+        return transaction {
+            let token = Token(storage: self)
+            observers[token] = action
+            action(.appended(requests))
+            return token
         }
-        
-        if let index = requests.firstIndex(where: { (req) -> Bool in
-            return request?.id == req.id ? true : false
-        }){
-            requests[index] = request!
-        }else{
-            requests.insert(request!, at: 0)
+    }
+
+    func saveRequest(request: RequestModel) {
+        transaction {
+            let knownRequestIndex = requests.lastIndex(where: { (req) -> Bool in
+                return request.id == req.id ? true : false
+            })
+
+            if let index = knownRequestIndex {
+                requests[index] = request
+                notify(.updated(request, at: index))
+            } else {
+                requests.append(request)
+                notify(.appended([request]))
+            }
+
+            if requests.count > 100 {
+                requests.removeFirst()
+                notify(.removed(indice: 0 ..< 1))
+            }
```

Comment:
Could it happen that `requests` get more than 101 items, e.g. if `observe` were called multiple times – enough to exceed that 100 limit – before this `saveRequest` gets called and clean them up? iow, should we check if we should drop more that just one request from this list?

Reply:
It is appended one by one so far, so it cannot spill above that for now.